src/util/newconfig/ConfigConstraints.cpp

```diff
@@ -0,0 +1,137 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2024, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include "util/newconfig/ConfigConstraints.hpp"
+
+#include "util/newconfig/Errors.hpp"
+
+#include <fmt/core.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <optional>
+#include <regex>
+#include <string>
+#include <string_view>
+#include <variant>
+
+namespace util::config {
+
+std::optional<Error>
+PortConstraint::checkConstraint(Constraint::ValueType const& port) const
+{
+    uint32_t p = 0;
+    if (!checkType(port))
+        return Error{"Port must be an integer or string"};
```

Comment:
To not duplicate everywhere this you could change a bit interface of `Constraint`:
```cpp
class Constraint {
public:
  std::optional<Error> check(value) {
     if (auto  maybeError = checkTypeImpl(value); maybeError.has_value())
        return std::move(maybeError).value();
     return checkValueImpl(value);
  }
protected:
  virtual std::optional<Error> checkTypeImpl(value) = 0;
  virtual std::optional<Error> checkValueImpl(value) = 0;
};
```